### ISSUE TITLE
amsmath \dots should do tests after expansion

### ISSUE THREAD

**Issue by davidcarlisle** [latex3/latex2e#1521]
As noted in #1448 `\dots` usues `\futurelet` testing to determine the next token.

this has lead to multiple additions in recent releases coping with `\protected` and classic `\protect` macros, `\long` macros and finally `\if..` tokens, It also has the issue that user-level macros are not expanded so hide any definition which would request a different dots form.

The suggestion is to "f-expand" (`\romannumeral`) the input stream before doing the tests.
this would expand user commands and allow the removal  of most of the `\meaning` tests checking for macro types.

`\DOTSB` etc so are all `\relax` so would be available for testing via a simple delimited macro if the futurelet token is ifx equal to `\relax`

Code for this wasn't quie ready for the 2024-10 release, but shoudl be made availabel vial a -dev release after that.

**Comment by peaR-red** [comment]
Here's my (rather hacky) attempt to reimplement `\dots` using `\romannumeral` expansion.
```latex
\def\mdots@{\@xp\mdots@a\romannumeral-`\0}
\def\mdots@a{\FN@\mdots@@}

\def\mdots@@{\gdef\thedots@{\dotso@}%
 \ifx\@let@token\boldsymbol
   \gdef\thedots@\boldsymbol{\boldsymboldots@}%
 \else
   \ifx,\@let@token \gdef\thedots@{\dotsc}%
   \else
     \ifx\not\@let@token
       \gdef\thedots@{\dotsb@}%
     \else
       \keybin@
       \ifgtest@
         \gdef\thedots@{\dotsb@}%
       \else
         \ifx\relax\@let@token
           \global\let\next@\mdots@relax
         \else
           \xdef\meaning@{\meaning\@let@token. .........}%
           \@xp\math@\meaning@\math@
           \ifgtest@
             \@xp\mathch@\meaning@\mathch@
             \ifgtest@
               \@xp\getmathch@\meaning@\getmathch@
             \else
               \@xp\mathbin@\meaning@\mathbin@
               \ifgtest@
                 \gdef\thedots@{\dotsb@}%
               \else
                 \@xp\mathrel@\meaning@\mathrel@
                 \ifgtest@
                   \gdef\thedots@{\dotsb@}%
                 \fi
               \fi
             \fi
           \else
             \@xp\Umathch@\meaning@"0"\Umathch@
             \ifgtest@
             \else
               \@xp\thecharacter@\meaning@\thecharacter@
             \fi
           \fi
           \global\let\next@\thedots@
         \fi %
       \fi %
     \fi %
   \fi %
 \fi
 \next@}

\def\mdots@@DOTSB#1\DOTSB#2\@nil{\ifx\@empty#2\@empty\else\gdef\thedots@{\dotsb@}\fi}
\def\mdots@@DOTSI#1\DOTSI#2\@nil{\ifx\@empty#2\@empty\else\gdef\thedots@{\dotsi}\fi}

\def\mdots@relax#1{%
  \edef\@tempa{\string#1}%
  \edef\@tempb{\string\protect}%
  \ifx\@tempa\@tempb
    \@xp\mdots@%
  \else
    \mdots@@DOTSB#1\DOTSB\@nil
    \mdots@@DOTSI#1\DOTSI\@nil
    \thedots@#1%
  \fi
}

\def\extra@{%
 \rightdelim@\ifgtest@
 \else\ifx\@let@token$\gtest@true
 \else\ifx\@let@token\DOTSX\gtest@true
 \fi\fi\fi}
```
This covers most reported cases (excluding explicit `\mathbin`, `\mathrel`, and `\(U)mathchar` tokens), including a bug that I stumbled upon the current `\dots`.

In the current `\dots` implementation, if one does, for example, `x \dots \boldsymbol{A + B}y`, then the internal `\@let@token` becomes `A`, and the remaining tokens (`+B`) are prematurely dumped directly into the execution stream twice in a row.

**Comment by FrankMittelbach** [comment]
> In the current `\dots` implementation, if one does, for example, `x \dots \boldsymbol{A + B}y`, then the internal `\@let@token` becomes `A`, and the remaining tokens (`+B`) are prematurely dumped directly into the execution stream twice in a row.

That sounds as if it need to be recorded as a separate issue including a MWE (even if a pull later closes both in one go)

**Comment by peaR-red** [comment]
> > In the current `\dots` implementation, if one does, for example, `x \dots \boldsymbol{A + B}y`, then the internal `\@let@token` becomes `A`, and the remaining tokens (`+B`) are prematurely dumped directly into the execution stream twice in a row.
> 
> That sounds as if it need to be recorded as a separate issue including a MWE (even if a pull later closes both in one go)

Yes, indeed. I actually intended to report it after commenting that, and after my current PR is merged.
